core/src/jsMain/kotlin/Observers.kt

```diff
@@ -1,129 +1,89 @@
-                addEventListener(CHARACTERISTIC_VALUE_CHANGED, observation.listener)
-                peripheral.ioLock.withLock {
-                    startNotifications().await()
-                }
-            }
+        .filter {
+            it.characteristic.characteristicUuid == characteristic.characteristicUuid
         }
-
-        try {
-            changes.collect {
-                if (it.characteristic.characteristicUuid == characteristic.characteristicUuid) {
-                    emit(it.data)
-                }
+        .map { it.data }
+        .onCompletion {
+            if (observations.remove(characteristic, onSubscription) < 1) {
+                peripheral.stopObservation(characteristic)
             }
-        } catch (t: Throwable) {
-            // Unnecessary `catch` block as workaround for KT-37279 (needed until we switch to IR compiler).
-            // https://youtrack.jetbrains.com/issue/KT-37279
-            // Once KT-37279 is fixed, this `catch` should be replaced with `finally`.
-            // See previous logic (before workaround) in:
-            // https://github.com/JuulLabs/kable/blob/151e54d255bf5595c67023927084d083e6180706/core/src/jsMain/kotlin/Observers.kt#L48-L72
-            observation.teardown(bluetoothRemoteGATTCharacteristic, characteristic)
-            return@flow
         }
-        observation.teardown(bluetoothRemoteGATTCharacteristic, characteristic)
-    }
-
-    private suspend fun Observation.teardown(
-        bluetoothRemoteGATTCharacteristic: BluetoothRemoteGATTCharacteristic,
-        characteristic: Characteristic
-    ) {
-        if (--count < 1) {
-            bluetoothRemoteGATTCharacteristic.apply {
-                /* Throws `DOMException` if connection is closed:
-                 *
-                 * DOMException: Failed to execute 'stopNotifications' on 'BluetoothRemoteGATTCharacteristic':
-                 * Characteristic with UUID [...] is no longer valid. Remember to retrieve the characteristic
-                 * again after reconnecting.
-                 *
-                 * Wrapped in `runCatching` to silently ignore failure, as notification will already be
-                 * invalidated due to the connection being closed.
-                 */
-                runCatching {
-                    peripheral.ioLock.withLock {
-                        stopNotifications().await()
-                    }
-                }
 
-                removeEventListener(CHARACTERISTIC_VALUE_CHANGED, listener)
-            }
-            observers.remove(characteristic)
+    suspend fun rewire() {
+        observations.entries.forEach { (characteristic, onSubscriptionActions) ->
+            peripheral.startObservation(characteristic)
+            onSubscriptionActions.forEach { it() }
         }
     }
+}
 
-    fun invalidate() {
-        observers.forEach { (_, observation) ->
-            observation.listener = null
+private class Observations {
```

Comment:
It's a little sad that this `Observations` class can't live in common code. I can see that they're different but it feels _super_ similar between the three implementations.

Reply:
I agree it's a little sad. Maybe one day I can figure out a way to make it generic and reusable, but for now, just getting it working. 🤷